scipy/stats/_stats_py.py

```diff
@@ -2849,8 +2853,11 @@ def obrientransform(*samples):
     if sLast:
         for arr in arrays[:-1]:
             if sLast != arr.shape:
-                return np.array(arrays, dtype=object)
-    return np.array(arrays)
+                if is_numpy(xp):
+                    return xp.array(arrays, dtype=object)
+                else:
+                    return arrays
```

Comment:
This doesn't seem super nice; alternative suggestions gratefully received

Reply:
Maybe we have the switch depend on whether `SCIPY_ARRAY_API=True` rather than whether `xp` is  NumPy? At some point I think we'll have to emit warnings before `SCIPY_ARRAY_API=1` becomes the default behavior; I don't think we will want to carry this object array return thing forward.

Reply:
Or maybe we deprecated the function and create an `obrien_transform` or `obrien` instead.

Reply:
Would it be too restrictive to require all the samples to have the same shape? Then this would be a fairly easy deprecation to remove

Reply:
Yes, it seems too restrictive because the transform doesn't seem to have that restriction. 

In fact, the transform on each array seems independent of the others. I was going to say that we could deprecate the use of multiple arguments, since there is no advantage to passing them in together. But it looks like then we'd always be adding an extra dimension to the result : /

Reply:
I see two options:

1. Create a new function and deprecate this one. If there are other breaking changes we would like to make then this could be a good idea, otherwise this is quite a lot of churn.
2. Add a `legacy` argument. Initially, this defaults to true, preserving the default behaviour but emitting a deprecation warning. Setting it to false triggers a new return-type behaviour. After two releases, the default changes from true to false, and then after another two releases, we remove it.

Reply:
I typically think option 1 is easier for us (not having to tiptoe around the old code) and just as easy or easier for users (e.g. add an underscore to the name rather than specifying a `legacy` argument and then having to remove it later). Option 1 is a lot of churn, but I tend to think the *two* deprecation cycles required by Option 2 is more churn than one cycle.

There are other options:
* Add `axis` and warn that specifying `axis` is required because the default is changing from `None` to `axis=0` to match other stats functions. Document that if the user specifies `axis`, the return type is a tuple rather than an array. At the end of the deprecation period, nothing needs to be removed, the return type is always a tuple, and the default `axis` is `0` instead of `None`.
* Defer deprecation to whenever `SCIPY_ARRAY_API=1` is going to become the only behavior. A lot of other things are going to need deprecation at that time. In the meantime, whether a tuple or object array is returned depends on `SCIPY_ARRAY_API`.

Reply:
> * Add `axis` and warn that specifying `axis` is required because the default is changing from `None` to `axis=0` to match other stats functions. Document that if the user specifies `axis`, the return type is a tuple rather than an array. At the end of the deprecation period, nothing needs to be removed, the return type is always a tuple, and the default `axis` is `0` instead of `None`.

This seems perhaps preferable as it kills multiple birds with one stone unless there are other things we would gain by rewriting.

If you are happy with this I can do this as a separate pr.

Reply:
Ok. We can try that, but it would need a discourse post. I can review the rest in the meantime.